File: 124-binary-tree-maximum-path-sum/124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        
        def dfs(root):
            if not root: return 0
            left = dfs(root.left)
            right = dfs(root.right)
            self.max = max(self.max, left+root.val, right+root.val, left+right+root.val, root.val)

            return max(root.val, left+root.val, right+root.val)
        
        self.max = -inf
        dfs(root)
        return self.max

File: 124-binary-tree-maximum-path-sum/124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = INT_MIN;
    int dfs(TreeNode* root) {
        if (root == NULL)
            return 0;
        int left = dfs(root->left);
        int right = dfs(root->right);
        
        int cur = root->val + max({left, right, 0});
        res = max({res, cur, left + right + root->val});

        return cur;
    }
    
    int maxPathSum(TreeNode* root) {
        dfs(root);
        return res;
    }
};